libcomposefs/lcfs-writer-erofs.c

```diff
@@ -1303,3 +1304,512 @@ int lcfs_write_erofs_to(struct lcfs_ctx_s *ctx)
+	struct erofs_inode_extended extended;
+} erofs_inode;
+
+static const erofs_inode *lcfs_image_get_erofs_inode(struct lcfs_image_data *data,
+						     uint64_t nid)
+{
+	const uint8_t *inode_data = data->erofs_metadata + (nid << EROFS_ISLOTBITS);
+
+	if (inode_data >= data->erofs_metadata_end)
+		return NULL;
+
+	return (const erofs_inode *)inode_data;
+}
+
+static uint16_t erofs_inode_version(const erofs_inode *cino)
+{
+	uint16_t i_format = lcfs_u16_from_file(cino->i_format);
+	return (i_format >> EROFS_I_VERSION_BIT) & EROFS_I_VERSION_MASK;
+}
+
+static bool erofs_inode_is_compact(const erofs_inode *cino)
+{
+	return erofs_inode_version(cino) == 0;
+}
+
+static uint16_t erofs_inode_datalayout(const erofs_inode *cino)
+{
+	uint16_t i_format = lcfs_u16_from_file(cino->i_format);
+	return (i_format >> EROFS_I_DATALAYOUT_BIT) & EROFS_I_DATALAYOUT_MASK;
+}
+
+static bool erofs_inode_is_tailpacked(const erofs_inode *cino)
+{
+	return erofs_inode_datalayout(cino) == EROFS_INODE_FLAT_INLINE;
+}
+
+static struct lcfs_node_s *lcfs_build_node_from_image(struct lcfs_image_data *data,
+						      uint64_t nid);
+
+static int erofs_readdir_block(struct lcfs_image_data *data,
+			       struct lcfs_node_s *parent, const uint8_t *block,
+			       size_t block_size)
+{
+	const struct erofs_dirent *dirents = (struct erofs_dirent *)block;
+	size_t dirents_size = lcfs_u16_from_file(dirents[0].nameoff);
+	size_t n_dirents, i;
+
+	if (dirents_size % sizeof(struct erofs_dirent) != 0) {
+		/* This should not happen for valid filesystems */
+		errno = EINVAL;
+		return -1;
+	}
+
+	n_dirents = dirents_size / sizeof(struct erofs_dirent);
+
+	for (i = 0; i < n_dirents; i++) {
+		char name_buf[PATH_MAX];
+		uint64_t nid = lcfs_u64_from_file(dirents[i].nid);
+		uint16_t nameoff = lcfs_u16_from_file(dirents[i].nameoff);
+		const char *child_name;
+		uint16_t child_name_len;
+		cleanup_node struct lcfs_node_s *child = NULL;
+
+		/* Compute length of the name, which is a bit weird for the last dirent */
+		child_name = (char *)(block + nameoff);
+		if (i + 1 < n_dirents)
+			child_name_len =
+				lcfs_u16_from_file(dirents[i + 1].nameoff) - nameoff;
+		else
+			child_name_len = strnlen(child_name, block_size - nameoff);
+
+		if ((child_name_len == 1 && child_name[0] == '.') ||
+		    (child_name_len == 2 && child_name[0] == '.' &&
+		     child_name[1] == '.'))
+			continue;
+
+		/* Copy to null terminate */
+		child_name_len = min(child_name_len, PATH_MAX - 1);
+		memcpy(name_buf, child_name, child_name_len);
+		name_buf[child_name_len] = 0;
+
+		child = lcfs_build_node_from_image(data, nid);
+		if (child == NULL)
+			return -1;
+
+		if ((child->inode.st_mode & S_IFMT) == S_IFCHR &&
+		    child->inode.st_rdev == 0)
+			continue; /* Skip whiteouts (we added) */
+
+		if (lcfs_node_add_child(parent, child, /* Takes ownership on success */
+					name_buf) < 0)
+			return -1;
+		steal_pointer(&child);
+	}
+
+	return 0;
+}
+
+const char *erofs_xattr_prefixes[] = {
```

Comment:
Surely it'd be trivial to share this in a `lcfs-internal-erofs.h` with the fuse code?

Reply:
yeah